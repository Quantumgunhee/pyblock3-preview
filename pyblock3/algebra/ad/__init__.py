
ENABLE_JAX = False
ENABLE_EINSUM = True