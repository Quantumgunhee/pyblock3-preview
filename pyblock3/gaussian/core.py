
import numpy as np
import torch


def pack(x):
    size = sum(z.nelement() for z in x)
    r = np.zeros(size)
    iz = 0
    for z in x:
        r[iz:iz + z.nelement()] = z.detach().numpy().flatten()
        iz += z.nelement()
    assert iz == r.size
    return r


def unpack(r, x):
    iz = 0
    for ix, z in enumerate(x):
        x[ix] = torch.tensor(
            r[iz:iz + z.nelement()].reshape(z.shape)).requires_grad_(True)
        iz += z.nelement()
    assert iz == r.size
    return x


class GaussianOptimizer:

    def __init__(self, tn, h1e, g2e, jit=True, iprint=1):
        self.tn = tn
        self.niter = 0

        def get_ener(x):
            tn.set_params(x)
            tn.unitize_()
            return tn.energy_tot(h1e, g2e)

        if jit:
            self.fn = torch.jit.trace(
                get_ener, example_inputs=[tn.get_params()])
        else:
            self.fn = get_ener

        def energy_and_grad(x):
            params = unpack(x, tn.get_params())
            energy = get_ener(params)
            grad = torch.autograd.grad(energy, params)
            self.niter += 1
            if iprint >= 1:
                print("%5d = %20.12f" % (self.niter, float(energy.detach())))
            return float(energy.detach()), pack(grad)

        self.grad = energy_and_grad

    def optimize(self, x0=None, maxiter=100):

        self.niter = 0

        if x0 is None:
            x0 = pack(self.tn.get_params())
        elif isinstance(x0, str) and x0 == "random":
            x0 = pack(self.tn.get_params())
            x0[:] = np.random.random(x0.shape)

        from scipy.optimize import minimize

        res = minimize(fun=self.grad, jac=True, hessp=None, x0=x0,
                       tol=None, method='L-BFGS-B', options={'maxiter': maxiter})

        return self.grad(res.x)[0], res.x


class GaussianTensor:
    """
    Attrs:
        u_idx : list(any), up (output) indices
        d_idx : list(any), down (output) indices
        data : square matrix, anti-symm K matrix for orbital rotation
        n_core : int, first n_core up indices have fixed occ (can be core or virtual)
        core_occ : list(float), the fixed occ (0 ~ 1)
    """

    def __init__(self, u_idx, d_idx, n_core=0, data=None):
        self.u_idx = u_idx
        self.d_idx = d_idx
        self.data = data
        self.n_core = n_core
        self.core_occ = None

    def __repr__(self):
        return "[u = %r d = %r nc = %r]" % (self.u_idx, self.d_idx, self.n_core)

    @property
    def width(self):
        return max(len(self.u_idx), len(self.d_idx))

    def get_du_map(self):
        return {i: j for i, j in zip(self.d_idx, self.u_idx)}

    def get_ud_map(self):
        return {i: j for i, j in zip(self.u_idx, self.d_idx)}


class GaussianTensorNetwork:

    def __init__(self, tensors):
        self.tensors = tensors

    def get_params(self):
        return [ts.data for ts in self.tensors]

    def set_params(self, params):
        for p, ts in zip(params, self.tensors):
            ts.data = p

    def unitize_(self):
        for ts in self.tensors:
            ts.data = 0.5 * (ts.data - ts.data.T)

    def get_initial_indices(self):
        """
        Initial indices are down (d_idx) indices that are not connected to any up indices.
        """

        all_d_idx = set()

        for ts in self.tensors:
            for di in ts.d_idx:
                all_d_idx.add(di)

        for ts in self.tensors:
            for ui in ts.u_idx:
                if ui in all_d_idx:
                    all_d_idx.remove(ui)

        return sorted(list(all_d_idx))

    def get_terminal_indices(self):
        """
        Terminal indices are up (d_idx) indices that are not connected to any down indices.
        """

        all_u_idx = set()

        for ts in self.tensors:
            for ui in ts.u_idx:
                all_u_idx.add(ui)

        for ts in self.tensors:
            for di in ts.d_idx:
                if di in all_u_idx:
                    all_u_idx.remove(di)

        return sorted(list(all_u_idx))

    def get_layers(self):
        """
        Return a list of layers, each layer is a list of indices of tensors
        that can be computed simultaneously, as long as all previous layers
        are computed.
        """

        current_idxs = set(self.get_initial_indices())
        layers = []
        current_tensor_idxs = range(0, len(self.tensors))

        while current_tensor_idxs != []:

            current_layer = []
            next_tensor_idxs = []
            next_idxs = current_idxs.copy()

            for its in current_tensor_idxs:
                ts = self.tensors[its]
                if all(ix in current_idxs for ix in ts.d_idx):
                    current_layer.append(its)
                    for ix in ts.d_idx:
                        next_idxs.remove(ix)
                    for ix in ts.u_idx:
                        next_idxs.add(ix)
                else:
                    next_tensor_idxs.append(its)

            current_idxs = next_idxs
            current_tensor_idxs = next_tensor_idxs

            layers.append(current_layer)

        return layers

    def __repr__(self):

        layers = self.get_layers()
        current_idxs = self.get_initial_indices()
        terminal_idxs = set(self.get_terminal_indices())

        pp = ['┆' * len(current_idxs)]

        for l in layers:
            pl = [' '] * len(current_idxs)
            layer_min_ip = min(min(current_idxs.index(xd) for xd in self.tensors[its].d_idx)
                               for its in l)
            layer_max_ip = max(max(current_idxs.index(xd) for xd in self.tensors[its].d_idx)
                               for its in l)
            for its in l:
                nx = self.tensors[its].width
                min_ip = min(current_idxs.index(xd)
                             for xd in self.tensors[its].d_idx)
                max_ip = max(current_idxs.index(xd)
                             for xd in self.tensors[its].d_idx)
                if len(l) == 1 or not (max_ip == layer_max_ip and min_ip == layer_min_ip):
                    pl[min_ip:max_ip] = '━' * (max_ip - min_ip)
                for ix, (xu, xd) in enumerate(zip(self.tensors[its].u_idx, self.tensors[its].d_idx)):
                    ip = current_idxs.index(xd)
                    current_idxs[ip] = xu
                    pl[ip] = '┃' if nx == 1 else (
                        '┣' if ix == 0 else ('┫' if ix == nx - 1 else '┿'))

            pp.append(pl)
            if l != layers[-1] or self.tensors[its].core_occ is not None:
                pl = [' '] * len(current_idxs)
                uset = set()
                for its in l:
                    uset |= set(self.tensors[its].u_idx)
                for ic, c in enumerate(current_idxs):
                    if c not in terminal_idxs:
                        if c in uset:
                            pl[ic] = '┆' or '|'
                        else:
                            pl[ic] = pp[-1][ic] = '┆'
                    elif self.tensors[its].core_occ is not None:
                        for its in l:
                            if c in self.tensors[its].u_idx:
                                iuc = self.tensors[its].u_idx.index(c)
                                occ = self.tensors[its].core_occ[iuc]
                                pl[ic] = str(int(np.round(occ)))
                pp.append(pl)

        return '\n'.join([''.join(x) for x in pp[::-1]])

    def fit_rdm1(self, dm, dm_idxs=None):

        if dm_idxs is None:
            dm_idxs = self.get_initial_indices()

        dm_idxs_map = {x: ix for ix, x in enumerate(dm_idxs)}
        dm = torch.clone(dm) / 2

        layers = self.get_layers()

        def my_logm(mrot):
            rs = mrot + mrot.T
            _, rv = torch.linalg.eigh(rs)
            rd = rv.T @ mrot @ rv
            ld = torch.zeros_like(rd)
            for i in range(0, len(rd) // 2 * 2, 2):
                xcos = (rd[i, i] + rd[i + 1, i + 1]) / 2
                xsin = (rd[i, i + 1] - rd[i + 1, i]) / 2
                theta = torch.arctan2(xsin, xcos)
                ld[i, i + 1] = theta
                ld[i + 1, i] = -theta
            return rv @ ld @ rv.T

        for l in layers:
            for its in l:
                # find the small 1pdm for indices at this tensor
                eff_dm_idxs = np.array([dm_idxs_map[x]
                                        for x in self.tensors[its].d_idx])
                eff_dm = dm[eff_dm_idxs, :][:, eff_dm_idxs]
                xb, ub = torch.linalg.eigh(eff_dm)
                # sort eigvals to put 0 and 1 near the beginning
                xg = torch.tensor([list(xb), list(1 - xb)])
                xg[xg <= 1E-50] = 1E-50
                p = torch.argsort(-(xg[0] * torch.log(xg[0]) +
                                    xg[1] * torch.log(xg[1])))
                xb = xb[p]
                ub = ub[:, p]
                if torch.linalg.det(ub) < 0:
                    ub[:, 0] *= -1
                self.tensors[its].data = my_logm(ub)
                self.tensors[its].core_occ = np.round(
                    xb[:self.tensors[its].n_core])
                dm[eff_dm_idxs, :] = ub.T @ dm[eff_dm_idxs, :]
                dm[:, eff_dm_idxs] = dm[:, eff_dm_idxs] @ ub
                # update index tags after apply this gate
                du_map = self.tensors[its].get_du_map()
                dm_idxs_map = {du_map.get(
                    x, x): ix for x, ix in dm_idxs_map.items()}

        return self

    def make_rdm1(self):

        dm_idxs = self.get_terminal_indices()
        dm_idxs_map = {x: ix for ix, x in enumerate(dm_idxs)}

        layers = self.get_layers()

        dm = torch.zeros((len(dm_idxs), len(dm_idxs)), dtype=torch.float64)

        for l in layers[::-1]:
            for its in l[::-1]:
                # find the small 1pdm for indices at this tensor
                eff_dm_idxs = np.array([dm_idxs_map[x]
                                        for x in self.tensors[its].u_idx])
                dmc_idxs = np.array([dm_idxs_map[u] for u
                                     in self.tensors[its].u_idx[:self.tensors[its].n_core]])
                dm[dmc_idxs, dmc_idxs] = self.tensors[its].core_occ
                ub = torch.matrix_exp(self.tensors[its].data)
                dm[eff_dm_idxs, :] = ub @ dm[eff_dm_idxs, :]
                dm[:, eff_dm_idxs] = dm[:, eff_dm_idxs] @ ub.T
                # update index tags after apply this gate
                ud_map = self.tensors[its].get_ud_map()
                dm_idxs_map = {ud_map.get(
                    x, x): ix for x, ix in dm_idxs_map.items()}

        init_idxs = self.get_initial_indices()
        reidx = np.array([dm_idxs_map[ix] for ix in init_idxs])
        return dm[reidx, :][:, reidx] * 2

    def make_rdm2(self, dm1=None):
        if dm1 is None:
            dm1 = self.make_rdm1()
        dm2 = (torch.einsum('ij,kl->ijkl', dm1, dm1)
               - torch.einsum('ij,kl->iklj', dm1, dm1) / 2)
        return dm2

    def energy_tot(self, h1e, g2e, dm1=None, dm2=None):
        if dm1 is None:
            dm1 = self.make_rdm1()
        if dm2 is None:
            dm2 = self.make_rdm2(dm1=dm1)
        e_tot = torch.einsum('ij,ij->', dm1, h1e)
        e_tot += torch.einsum('ijkl,ijkl->', dm2, g2e) * 0.5
        return e_tot

    def get_occupations(self):

        term_occ = []

        layers = self.get_layers()

        for l in layers:
            for its in l:
                for occ in self.tensors[its].core_occ:
                    term_occ.append(occ)

        return np.array(term_occ) * 2

    def set_occ_smearing(self, sigma=0.0):

        layers = self.get_layers()

        for l in layers:
            for its in l:
                occs = self.tensors[its].core_occ
                if len(occs) > 0:
                    occs = torch.round(occs)
                    occs[occs > 0.5] -= sigma
                    occs[occs < 0.5] += sigma
                    self.tensors[its].core_occ = occs

    def set_occ_half_filling(self):

        layers = self.get_layers()

        ix = 0
        for l in layers:
            for its in l:
                occs = self.tensors[its].core_occ
                for i in range(len(occs)):
                    occs[i] = 1.0 if ix % 2 else 0.0
                    ix += 1
        
        return self


class GaussianMERA1D(GaussianTensorNetwork):

    def __init__(self, n_sites, n_tensor_sites=5, n_core=2, dis_ent=True, periodic=True):
        """
        Args:
            n_sites : number of sites in the MERA
            n_tensor_sites : number of sites in each MERA tensor
            n_core : number of core indices in the upper bond of MERA tensor
            dis_ent : if true, add disentangler layers
            periodic : if true, add extra non-local disentangler tensor for boundary sites
        """
        assert n_core >= 1
        tensors = []
        cur_idx = list(range(0, n_sites))
        idx = n_sites
        n_act = n_tensor_sites - n_core
        while True:
            if dis_ent:
                if not periodic:
                    next_idx = cur_idx[:n_tensor_sites // 2]
                    i = n_tensor_sites // 2
                    while True:
                        d = min(n_tensor_sites, len(cur_idx) - i)
                        d_idx = cur_idx[i:i + d]
                        u_idx = list(range(idx, idx + d))
                        if d == n_tensor_sites:
                            idx += d
                            tensors.append(GaussianTensor(
                                u_idx, d_idx, n_core=0))
                            next_idx += u_idx
                        else:
                            next_idx += d_idx
                        if i + d == len(cur_idx):
                            break
                        i += d
                else:
                    next_idx = []
                    i = n_tensor_sites // 2
                    while True:
                        d = min(n_tensor_sites, len(cur_idx) -
                                i + n_tensor_sites // 2)
                        d_idx = cur_idx[i:i + d]
                        u_idx = list(range(idx, idx + d))
                        if i + d > len(cur_idx):
                            if d == n_tensor_sites and d != len(cur_idx):
                                idx += d
                                tensors.append(GaussianTensor(
                                    u_idx[:len(d_idx)] + u_idx[len(d_idx):],
                                    d_idx + cur_idx[:i + d - len(cur_idx)],
                                    n_core=0))
                                next_idx = u_idx[len(
                                    d_idx):] + next_idx + u_idx[:len(d_idx)]
                            else:
                                next_idx = cur_idx[:i + d -
                                                   len(cur_idx)] + next_idx + d_idx
                        else:
                            if d == n_tensor_sites and d != len(cur_idx):
                                idx += d
                                tensors.append(GaussianTensor(
                                    u_idx, d_idx, n_core=0))
                                next_idx += u_idx
                            else:
                                next_idx += d_idx
                        if i + d == len(cur_idx) + n_tensor_sites // 2:
                            break
                        i += d
                cur_idx = next_idx
            next_idx = []
            i = 0
            while True:
                d = min(n_tensor_sites, len(cur_idx) - i)
                d_idx = cur_idx[i:i + d]
                u_idx = list(range(idx, idx + d))
                idx += d
                tensors.append(GaussianTensor(u_idx, d_idx, n_core=n_core))
                tensors[-1].n_core = max(0, tensors[-1].width - n_act)
                next_idx += u_idx[tensors[-1].n_core:]
                if i + d == len(cur_idx):
                    break
                i += d
            if tensors[-1].width == len(cur_idx):
                tensors[-1].n_core = tensors[-1].width
                break
            cur_idx = next_idx

        super().__init__(tensors)


class GaussianMPS(GaussianTensorNetwork):

    def __init__(self, n_sites, n_tensor_sites=5, n_core=2):
        """
        Args:
            n_sites : number of sites in the MPS
            n_tensor_sites : number of sites in each MPS tensor
            n_core : number of core indices in the upper bond of MPS tensor
        """
        assert n_core >= 1
        tensors = []
        cur_idx = list(range(0, n_sites))
        idx = n_sites
        i = 0
        while True:
            d = min(n_tensor_sites, n_sites - i)
            d_idx = cur_idx[i:i + d]
            u_idx = list(range(idx, idx + d))
            idx += d
            cur_idx[i:i + d] = u_idx
            tensors.append(GaussianTensor(u_idx, d_idx, n_core=n_core))
            if i + d == n_sites:
                tensors[-1].n_core = tensors[-1].width
                break
            i += n_core

        super().__init__(tensors)
