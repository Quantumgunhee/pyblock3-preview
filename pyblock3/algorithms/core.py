
import numpy as np
import pyblock3.algebra.funcs as pbalg
from enum import Enum, auto


def fmt_size(i, suffix='B'):
    if i < 1000:
        return "%d %s" % (i, suffix)
    else:
        a = 1024
        for pf in "KMGTPEZY":
            p = 2
            for k in [10, 100, 1000]:
                if i < k * a:
                    return "%%.%df %%s%%s" % p % (i / a, pf, suffix)
                p -= 1
            a *= 1024
    return "??? " + suffix

class DecompositionTypes(Enum):
    DensityMatrix = auto()
    SVD = auto()


class NoiseTypes(Enum):
    Perturbative = auto()
    Random = auto()


class SweepAlgorithm:
    def __init__(self, cutoff=1E-14,
                 decomp_type=DecompositionTypes.DensityMatrix,
                 noise_type=NoiseTypes.Perturbative):
        self.cutoff = cutoff
        self.decomp_type = decomp_type
        self.noise_type = noise_type

    def add_dm_noise(self, dm, mpo, wfn, noise, forward):
        if noise == 0:
            return dm
        if self.noise_type == NoiseTypes.Random:
            pdm = dm.__class__.random_like(dm)
            return dm + (noise / np.linalg.norm(pdm)) * pdm
        elif self.noise_type == NoiseTypes.Perturbative:
            if forward:
                pket = np.tensordot(
                    wfn[0], mpo[0], axes=((1, 2), (3, 4)))
                pket.normalize_along_axis(axis=7)
                pdm = np.tensordot(pket, pket, axes=((1, 2, 3, 4, 7), ) * 2)
            else:
                pket = np.tensordot(
                    mpo[1], wfn[0], axes=((3, 4), (3, 4)))
                pket.normalize_along_axis(axis=0)
                pdm = np.tensordot(pket, pket, axes=((0, 3, 4, 5, 6), ) * 2)
            norm = np.linalg.norm(pdm)
            return dm + (noise / norm) * pdm if abs(norm) > 1E-12 else dm
        else:
            assert False

    def add_wfn_noise(self, wfn, noise, forward):
        if noise == 0:
            return wfn
        if self.noise_type == NoiseTypes.Random:
            pwfn = wfn.__class__.random_like(wfn)
            return wfn + (np.sqrt(noise) / np.linalg.norm(pwfn)) * pwfn
        else:
            assert False

    def decomp_two_site(self, mpo, wfns, forward, noise, bond_dim, weights=None):
        wfn = wfns[0] if isinstance(wfns, list) else wfns
        if weights is None:
            weights = [1.0 / len(wfns)] * len(wfns)
        if self.decomp_type == DecompositionTypes.DensityMatrix:
            if forward:
                dm = np.tensordot(
                    wfn[0], wfn[0], axes=((-3, -2, -1), ) * 2)
                if isinstance(wfns, list):
                    dm = dm * weights[0]
                    for ex_wfn, w in zip(wfns[1:], weights[1:]):
                        dm = dm + w * np.tensordot(
                            ex_wfn[0], ex_wfn[0], axes=((-3, -2, -1), ) * 2)
                dm = self.add_dm_noise(dm, mpo, wfn, noise, forward)
                lsr = dm.tensor_svd(idx=3, pattern='++++++')
                l, _, _, error = pbalg.truncate_svd(
                    *lsr, cutoff=self.cutoff, max_bond_dim=bond_dim, eigen_values=True)
                wfn[:] = [l, np.tensordot(
                    l, wfn[0], axes=((0, 1, 2), ) * 2)]
            else:
                dm = np.tensordot(
                    wfn[0], wfn[0], axes=((0, 1, 2), ) * 2)
                if isinstance(wfns, list):
                    dm = dm * weights[0]
                    for ex_wfn, w in zip(wfns[1:], weights[1:]):
                        dm = dm + w * np.tensordot(
                            ex_wfn[0], ex_wfn[0], axes=((0, 1, 2), ) * 2)
                dm = self.add_dm_noise(dm, mpo, wfn, noise, forward)
                lsr = dm.tensor_svd(idx=3, pattern='-+--+-')
                _, _, r, error = pbalg.truncate_svd(
                    *lsr, cutoff=self.cutoff, max_bond_dim=bond_dim, eigen_values=True)
                wfn[:] = [np.tensordot(
                    wfn[0], r, axes=((-3, -2, -1), ) * 2), r]
        elif self.decomp_type == DecompositionTypes.SVD:
            assert not isinstance(wfns, list)
            wfn = self.add_wfn_noise(
                wfn[0], noise, forward)
            lsr = wfn.tensor_svd(idx=3, pattern='+++-+-')
            l, s, r, error = pbalg.truncate_svd(
                *lsr, cutoff=self.cutoff, max_bond_dim=bond_dim)
            if noise == 0:
                if forward:
                    wfn[:] = [l, np.tensordot(
                        s.diag(), r, axes=1)]
                else:
                    wfn[:] = [np.tensordot(
                        l, s.diag(), axes=1), r]
            else:
                if forward:
                    wfn[:] = [l, np.tensordot(
                        l, wfn[0], axes=((0, 1, 2), ) * 2)]
                else:
                    wfn[:] = [np.tensordot(
                        wfn[0], r, axes=((-3, -2, -1), ) * 2), r]
        else:
            assert False
        return error
