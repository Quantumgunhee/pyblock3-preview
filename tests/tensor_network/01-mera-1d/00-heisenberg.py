
# 1D Heisenberg model with 8 sites
# E dmrg = -3.374932598651

import numpy as np
import quimb.tensor as qtn
import warnings
import torch
import time
from pyblock3.heisenberg import Heisenberg

torch.autograd.set_detect_anomaly(False)
torch.set_num_threads(28)

import pyblock3.algebra.ad
pyblock3.algebra.ad.ENABLE_JAX = True
pyblock3.algebra.ad.ENABLE_AUTORAY = True

from pyblock3.algebra.ad import core
core.ENABLE_FUSED_IMPLS = True

from pyblock3.algebra.ad.core import SparseTensor

warnings.filterwarnings(action='ignore', category=torch.jit.TracerWarning)

import autoray as ar

SparseTensor.shape = property(lambda x: tuple(ix.n_bonds for ix in x.infos))
ar.register_function('pyblock3', 'array', lambda x: x)
ar.register_function('pyblock3', 'to_numpy', lambda x: x)
ar.register_function('torch', 'conjugate', torch.conj)

def torch_split_wrapper(old_fn):
    def new_fn(ary, indices_or_sections, *args, **kwargs):
        if len(indices_or_sections) == 0:
            return [ary]
        else:
            return old_fn(ary, indices_or_sections, *args, **kwargs)
    return new_fn

ar.register_function('torch', 'split', torch_split_wrapper, wrap=True)

L = 8
D = 20

np.random.seed(1234)

print(core.ENABLE_FUSED_IMPLS)

topology = [(i, i + 1, 1.0) for i in range(L - 1)]
hamil = Heisenberg(twos=1, n_sites=L, topology=topology, flat=False)

class MERAState:

    @staticmethod
    def get_disentanglers_and_isometries(basis, target, low=0, high=1, dtype=float, max_bond=None):
        tensors_uni = []
        tensors_iso = []
        tensors_cap = []
        while len(basis) > 1:
            n_sites = len(basis) // 2
            uni_basis = []
            for i in range(n_sites):
                info_high = basis[i * 2] | basis[i * 2 + 1]
                if max_bond is not None:
                    info_high.truncate_no_keep(max_bond)
                ts = SparseTensor.random(
                    (basis[i * 2], basis[i * 2 + 1], info_high, info_high),
                    dtype=dtype, pattern="++--") * (high - low) + low
                tensors_uni.append(ts)
                uni_basis.append(info_high)

            iso_basis = []
            for i in range(n_sites):
                info_high = uni_basis[i] * uni_basis[(i + 1) % n_sites]
                if max_bond is not None:
                    info_high.truncate_no_keep(max_bond)
                if n_sites == 1:
                    info_high = info_high.filter(info_high.__class__({target : 1}))
                ts = SparseTensor.random(
                    (uni_basis[i], uni_basis[(i + 1) % n_sites], info_high),
                    dtype=dtype, pattern="++-") * (high - low) + low
                if n_sites == 1:
                    tensors_cap.append(ts)
                else:
                    tensors_iso.append(ts)
                iso_basis.append(info_high)

            basis = iso_basis

        return tensors_uni, tensors_iso, tensors_cap

    @staticmethod
    def unitize_(mera):
        for t in mera:
            if t.data.ndim - len(t.left_inds) == 1:
                new_data = t.data.left_canonicalize()[0]
            elif t.data.ndim - len(t.left_inds) == 2:
                fuse_info = t.data.kron_sum_info(2, 3, pattern=t.data.pattern[2:])
                new_data = t.data.fuse(2, 3, info=fuse_info).left_canonicalize()[0].unfuse(2, info=fuse_info)
                new_data.pattern = t.data.pattern
            else:
                assert False
            t.modify(data=new_data)
        return mera

    @staticmethod
    def sp_operator(tj, basis, isite, dtype=float):
        Q = basis[isite].symm_class
        repr = SparseTensor.zeros(bond_infos=(basis[isite], basis[isite]), dtype=dtype, dq=Q(0, 2, 0), pattern="+-")
        for tm in range(-tj, tj - 1, 2):
            repr[Q(0, tm, 0)].data[0, 0] = np.sqrt((tj - tm) * (tj + tm + 2) // 4)
        return repr

    @staticmethod
    def sm_operator(tj, basis, isite, dtype=float):
        Q = basis[isite].symm_class
        repr = SparseTensor.zeros(bond_infos=(basis[isite], basis[isite]), dtype=dtype, dq=Q(0, -2, 0), pattern="+-")
        for tm in range(-tj + 2, tj + 1, 2):
            repr[Q(0, tm, 0)].data[0, 0] = np.sqrt((tj + tm) * (tj - tm + 2) // 4)
        return repr

    @staticmethod
    def sz_operator(tj, basis, isite, dtype=float):
        Q = basis[isite].symm_class
        repr = SparseTensor.zeros(bond_infos=(basis[isite], basis[isite]), dtype=dtype, dq=Q(0, 0, 0), pattern="+-")
        for tm in range(-tj, tj + 1, 2):
            repr[Q(0, tm, 0)].data[0, 0] = tm / 2.0
        return repr

    @staticmethod
    def get_hamil_terms(tj, basis, topology, dtype=float):
        terms = []
        sp_op = [None] * len(basis)
        sm_op = [None] * len(basis)
        sz_op = [None] * len(basis)
        for m in range(len(basis)):
            sp_op[m] = MERAState.sp_operator(tj, basis, m, dtype=dtype)
            sm_op[m] = MERAState.sm_operator(tj, basis, m, dtype=dtype)
            sz_op[m] = MERAState.sz_operator(tj, basis, m, dtype=dtype)

        for (i, j, v) in topology:
            assert i != j
            ts = sum([
                np.einsum('ij,kl->ikjl', 0.5 * v * sp_op[i], sm_op[j]),
                np.einsum('ij,kl->ikjl', 0.5 * v * sm_op[i], sp_op[j]),
                np.einsum('ij,kl->ikjl', v * sz_op[i], sz_op[j])
            ])
            terms.append(((i, j), ts))

        return terms


class TensorBuilder:
    def __init__(self, hamil, max_bond=None):
        self.uni, self.iso, self.cap = MERAState.get_disentanglers_and_isometries(hamil.basis, hamil.target, max_bond=max_bond)
        self.i_uni = 0
        self.i_iso = 0
        self.i_cap = 0

    @property
    def uni_fill_fn(self):
        def uni_fill_fn(shape):
            uni = self.uni[self.i_uni]
            self.i_uni += 1
            print('uni ', uni.shape, shape)
            assert uni.shape == shape
            return uni
        return uni_fill_fn

    @property
    def iso_fill_fn(self):
        def iso_fill_fn(shape):
            iso = self.iso[self.i_iso]
            self.i_iso += 1
            print('iso ', iso.shape, shape)
            # assert iso.shape == shape
            return iso
        return iso_fill_fn

    @property
    def cap_fill_fn(self):
        def cap_fill_fn(shape):
            cap = self.cap[self.i_cap]
            self.i_cap += 1
            print('cap ', cap.shape, shape)
            # assert cap.shape == shape
            return cap
        return cap_fill_fn

def get_params(self):
    from jax.tree_util import tree_flatten
    params, tree = tree_flatten(self)
    self._tree = tree
    return params

def set_params(self, params):
    from jax.tree_util import tree_unflatten
    x = tree_unflatten(self._tree, params)
    self.blocks = x.blocks
    self.pattern = x.pattern

SparseTensor.get_params = get_params
SparseTensor.set_params = set_params

tensor_builder = TensorBuilder(hamil, max_bond=D)

from quimb.experimental.merabuilder import MERA

mera = MERA.from_fill_fn(
    fill_fn=None,
    uni_fill_fn=tensor_builder.uni_fill_fn,
    iso_fill_fn=tensor_builder.iso_fill_fn,
    cap_fill_fn=tensor_builder.cap_fill_fn,
    L=8, D=D, block_size=2, cyclic=True
)

h_terms = MERAState.get_hamil_terms(hamil.twos, hamil.basis, topology)
terms = { k : v for k, v in h_terms }

def norm_fn(mera):
    return MERAState.unitize_(mera.copy())

def local_expectation(mera, terms, where, optimize='auto-hq'):
    tags = [mera.site_tag(coo) for coo in where]
    mera_ij = mera.select(tags, 'any')
    mera_ij_G = mera_ij.gate(terms[where], where)
    mera_ij_ex = (mera_ij_G & mera_ij.H)
    return mera_ij_ex.contract(all, optimize=optimize)

def loss_fn(mera, terms, **kwargs):
    with qtn.contraction.contract_backend("numpy"):
        r = sum(
            local_expectation(mera, terms, where, **kwargs)
            for where in terms
        )
        return r

mera = norm_fn(mera)
print(loss_fn(mera, terms, optimize='auto-hq'))

def _tn_callback(self):
    if not hasattr(self, 'tt'):
        self.it = 0
        self.tt = time.perf_counter()
    self.loss_best = min(self.loss_best, self.loss)
    tt = time.perf_counter() - self.tt
    self.tt = time.perf_counter()
    self.it += 1
    msg = f"{self.it:5d} {self.loss:+.12f} [best: {self.loss_best:+.12f}] T = {tt:10.3f}"
    print(msg)

tnopt = qtn.TNOptimizer(
    mera,
    loss_fn=loss_fn,
    norm_fn=norm_fn,
    loss_constants={'terms': terms},
    loss_kwargs={'optimize': 'auto-hq'},
    autodiff_backend='torch',
    callback=_tn_callback,
    progbar=False,
    jit_fn=False,
)

tt = time.perf_counter()
tnopt.optimizer = 'l-bfgs-b'
tnopt.optimize(500)
# tnopt.optimizer = 'adam'
# mera_opt = tnopt.optimize(1000)

print(min(tnopt.losses))
print('time = ', time.perf_counter() - tt)
