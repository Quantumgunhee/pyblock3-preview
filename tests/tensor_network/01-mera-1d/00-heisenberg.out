True
uni  (2, 2, 2, 2) (2, 2, 2, 2)
uni  (2, 2, 2, 2) (2, 2, 2, 2)
uni  (2, 2, 2, 2) (2, 2, 2, 2)
uni  (2, 2, 2, 2) (2, 2, 2, 2)
iso  (2, 2, 4) (2, 2, 4)
iso  (2, 2, 4) (2, 2, 4)
iso  (2, 2, 4) (2, 2, 4)
iso  (2, 2, 4) (2, 2, 4)
uni  (4, 4, 4, 4) (4, 4, 4, 4)
uni  (4, 4, 4, 4) (4, 4, 4, 4)
iso  (4, 4, 16) (4, 4, 16)
iso  (4, 4, 16) (4, 4, 16)
cap  (16, 16, 1) (16, 16)
-0.3222115365795482
    1 -0.322211536580 [best: -0.322211536580] T =      0.000
    2 -1.667873174867 [best: -1.667873174867] T =      1.133
    3 -2.384266913825 [best: -2.384266913825] T =      1.128
    4 -2.669988241652 [best: -2.669988241652] T =      1.128
    5 -3.132118831676 [best: -3.132118831676] T =      1.113
    6 -3.219837022153 [best: -3.219837022153] T =      1.126
    7 -3.283755655398 [best: -3.283755655398] T =      1.020
    8 -3.322141019995 [best: -3.322141019995] T =      1.134
    9 -3.342020434399 [best: -3.342020434399] T =      1.134
   10 -3.355227285436 [best: -3.355227285436] T =      1.148
   11 -3.366615703683 [best: -3.366615703683] T =      1.120
   12 -3.370101171668 [best: -3.370101171668] T =      1.135
   13 -3.373017050942 [best: -3.373017050942] T =      1.127
   14 -3.373864303933 [best: -3.373864303933] T =      1.399
   15 -3.374424161335 [best: -3.374424161335] T =      1.149
   16 -3.374831416102 [best: -3.374831416102] T =      1.122
   17 -3.374731987058 [best: -3.374831416102] T =      1.129
   18 -3.374884393618 [best: -3.374884393618] T =      1.121
   19 -3.374915900865 [best: -3.374915900865] T =      1.218
   20 -3.374925419978 [best: -3.374925419978] T =      1.134
   21 -3.374931410337 [best: -3.374931410337] T =      1.042
   22 -3.374931381846 [best: -3.374931410337] T =      1.117
   23 -3.374932290547 [best: -3.374932290547] T =      1.124
   24 -3.374932492227 [best: -3.374932492227] T =      1.135
   25 -3.374932567121 [best: -3.374932567121] T =      1.130
   26 -3.374932592222 [best: -3.374932592222] T =      1.140
   27 -3.374932592896 [best: -3.374932592896] T =      1.138
   28 -3.374932596038 [best: -3.374932596038] T =      1.149
-3.3749325960376293
time =  31.92972065322101
